named tensor support for torch.equal

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29322 named tensor support for torch.equal**
* #29307 Make TensorName::unifyFromRight in-place for efficiency

torch.equal checks if two tensors are equal in both size and values. For
named tensors, it also checks that the names are exactly equal. There is
an argument to be made for alternative semantics (check that the names
*match*), but for an API that is called "equal" I would expect it to
check equality on names as well.

Test Plan:
- new tests

Differential Revision: [D18453387](https://our.internmc.facebook.com/intern/diff/D18453387)

zou3519 committed:
named tensor support for torch.equal

torch.equal checks if two tensors are equal in both size and values. For
named tensors, it also checks that the names are exactly equal. There is
an argument to be made for alternative semantics (check that the names
*match*), but for an API that is called "equal" I would expect it to
check equality on names as well.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "named tensor support for torch.equal"


torch.equal checks if two tensors are equal in both size and values. For
named tensors, it also checks that the names are exactly equal. There is
an argument to be made for alternative semantics (check that the names
*match*), but for an API that is called "equal" I would expect it to
check equality on names as well.

Test Plan:
- new tests

[ghstack-poisoned]

nairbv commented on aten/src/TH/generic/THTensorMoreMath.cpp:
```diff
@@ -42,6 +40,16 @@ int THTensor_(equal)(THTensor *ta, THTensor* tb)
   return equal;
 }
 
+int THTensor_(equal)(THTensor *ta, THTensor* tb) {
```
it looks like this is only called from functions like `bool _th_equal` in LegacyTHFunctionsCPU.cpp. Would it be easy to return a bool here (and in the cuda version) instead of returning an int and casting farther up the stack?

zou3519 replied:
The original function returned an `int` so I am following that convention. I don't know the answer to whether or not a bool return here changes the behavior.

nairbv commented on aten/src/ATen/NamedTensorUtils.cpp:
```diff
@@ -513,6 +513,13 @@ std::vector<Dimname> compute_baddbmm_outnames(
 }
 
+bool are_names_equal(TensorImpl* self, TensorImpl* other) {
+  if (!impl::has_names(self) && !impl::has_names(other)) {
```
nit: I don't know if this is better since it might be less clear to read, but I'd be tempted to put:
```
if (!impl::has_names(self)) {
  return !impl::has_names(other);
}
```
short-circuits the case where they're not equal and only the second arg has names.

zou3519 replied:
That's not correct. If `other` doesn't have names and `self` does have names, their names can still be equal if all of `self`'s names are None.

nairbv reviewed (approved):
minor suggestions but looks good

zou3519 committed:
Update on "named tensor support for torch.equal"


torch.equal checks if two tensors are equal in both size and values. For
named tensors, it also checks that the names are exactly equal. There is
an argument to be made for alternative semantics (check that the names
*match*), but for an API that is called "equal" I would expect it to
check equality on names as well.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "named tensor support for torch.equal"


torch.equal checks if two tensors are equal in both size and values. For
named tensors, it also checks that the names are exactly equal. There is
an argument to be made for alternative semantics (check that the names
*match*), but for an API that is called "equal" I would expect it to
check equality on names as well.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "named tensor support for torch.equal"


torch.equal checks if two tensors are equal in both size and values. For
named tensors, it also checks that the names are exactly equal. There is
an argument to be made for alternative semantics (check that the names
*match*), but for an API that is called "equal" I would expect it to
check equality on names as well.

Test Plan:
- new tests

[ghstack-poisoned]
